### ISSUE TITLE
Canon Powershot A560 Adobe Dng strange colors

### ISSUE THREAD

**Issue by falket** [Beep6581/RawTherapee#5147]
I’m trying to develop a picture i took with a Canon Powershot A560 (not mine), Raw Therapee 5.5 release.

With the original chdk’s dng i get right colors, with a dng converted with Adobe Dng Converter (in order to get a lossless compression and a better preview) colors are completely off.

The original dng is on raw.pixls.us -> https://raw.pixls.us/getfile.php/2918/nice/Canon%20-%20Canon%20PowerShot%20A560%20-%2010bit%20CHDK%20ver.%201.4.1%20(4:3).DNG

The compressed one together @pixls.us here: https://pixls-discuss.s3.dualstack.us-east-1.amazonaws.com/original/3X/b/9/b94c03126e7e868d778936e38a20266c1912d702.DNG

Sample https://discuss.pixls.us/t/canon-powershot-a560-adobe-dng-strange-colors/10768

Thank you in adv.

Version: 5.5
Branch: releases
Commit: d9ac235
Commit date: 
Compiler: cc 7.3.1
Processor: x86_64
System: Linux
Bit depth: 64 bits
Gtkmm: V3.22.2
Lensfun: V0.3.2.0
Build type: release
Build flags:  -std=c++11  -Werror=unused-label -Wall -Wuninitialized -Wno-deprecated-declarations -Wno-unused-result -fopenmp -Werror=unknown-pragmas -O2 -DNDEBUG -ftree-vectorize
Link flags:   
OpenMP support: ON
MMAP support: ON

**Comment by iliasg** [comment]
For some reason RT translates wrongly the color matrix of Adobe's DNG although the matrix looks OK  .. 

Console reports very strange coefficients
`rgb_cam:[ [ 3.155447 2.023462 -2.332113], [5.648332 -0.065382 -0.038343], [-7.803779 -0.958081 3.370456] ]`

Strange that RT translates correctly the CHDK matrix although it's not in a standard format but an integer one !! 
`rgb_cam:[ [ 1.362567 -0.175834 0.029739], [-0.186725 1.547345 -0.461461], [-0.175842 -0.371511 1.431722] ]`

@heckflosse Ingo, Is there special code active for CHDKs only ??

**Comment by agriggio** [comment]
@iliasg we have a special case not for CHDK, but for the Adobe DNG converter: basically, in order to ensure that we get the "same colours" between an original raw and its DNG conversion, we ignore the color matrices for Adobe DNG. In this case, we don't have a camconst entry for the camera, which I suppose it's the reason why we get weird results. Here's a suitable one, derived from the one in the Adobe DNG (Bradford adatped from StdA to D65):

```txt
    {
        "make_model": [ "Canon PowerShot A560" ],
        "dcraw_matrix": [ 13300, -6137,  518, -6895, 14734, 3171, -1080, 3167, 16605]
    },
```

**Comment by iliasg** [comment]
@agriggio Alberto, I had the impression that RT ignores the Adobe_DNG's colormatrices only if we have a camconst item for the specific model.

If no camconst entry exists then RT should use the available DNG colormatrix .. and I think it indeed makes use** of the colormatrix1 data .. but using a wrong interpetation. And it's not because it is tagged as a "Standard Light A" illuminant because even after editing the data using your bradford adapted D65 colormatrix the rendering is false .. 

** If I choose "no profile" in RT the colors come almost fine ;)

**Comment by agriggio** [comment]
I might remember wrong, that's certainly possible...
but what do you mean by "editing the data?" you changed the matrix in the dng?

**Comment by iliasg** [comment]
Yes, changed the colormatrix and illuminant type with exiftool + exiftoolgui

**Comment by agriggio** [comment]
ok, I'll have a look

**Comment by agriggio** [comment]
@iliasg I think it works as I wrote previously, for the DNG converter we currently never use the embedded matrix. That's why editing it with exiftool has no effect.

**Comment by iliasg** [comment]
@agriggio But this (never using the colormatrix from AdobeDNG's conversions) is wrong .. destroys compatibility of many models for which we have no dcraw.cc nor camconst support ..

**Comment by iliasg** [comment]
@agriggio Alberto .. something is going wrong with DNGs
https://github.com/Beep6581/RawTherapee/issues/5126
There although we have camconst item for Leica SL RT uses the embedded matrix for the in_camera DNG .. colors are almost same but the WB calculations fail ..

**Comment by agriggio** [comment]
Ilias, last time I checked the code in RT gives priority to the embedded matrix over camconst, *for all dngs except those produced by Adobe dng converter*. this last part was a special case that I added, to fix the problem with colour mismatches between dng and original raw. so what you say is not surprising. I agree it would be nice to have more robust handling of the embedded matrix, and to add adaptation when a d65 one is not available. but it's not something I'm volunteering for, sorry...

**Comment by iliasg** [comment]
@agriggio Alberto, for Leica SL and other Leica DNGs and Pentax DNGs we use "adobe_coeff"  in Dcraw.cc for this exact reason (color matrix consistency) .. are these lines deactivated by a later code ?.

https://github.com/Beep6581/RawTherapee/blob/4fec856fbae6f4565c48b3ce8b7b7de5a8582ed5/rtengine/dcraw.cc#L10184

**Comment by agriggio** [comment]
@iliasg not that I'm aware of, but there might be a bug. I'll try to find some time to check
